### PR TITLE
Feature/implementing domain drains

### PR DESCRIPTION
**Pull request by davidporter-id-au**
Refactors the frontend Workflow and Admin handlers to take DomainHandler as a common parameter rather than doing the weird self-pointing thing they were doing in their constructors previously. 

Refactors the isolation-group implementation by splitting the single package into three parts: 

- The base `isolationgroup` package which just specifies the state interface and common types
- A default implementation of the state package as a subpackage so it can be split up a bit into smaller files
- A handler subpackage which is responsible for doing CRUD operations for frontend and mapping.

Refactors config store:
- Compose the dynamicConfig and Daemon into a unified interface for the config store to represent the fact that it's a stateful implementation of the dynamic config client.
- To remove the `done` channel and instead use a daemon interface
- Adds requisite Start/Stop functions

Refactors the unit-test for adminHandler_test.go 
- Mocks out the isolationgroup handler rather than mocking below

### CODE REVIEW

File: service/frontend/adminHandler.go [hunk 1]
```diff
@@ -142,44 +145,46 @@ var (
 	}
 )
 
-// NewAdminHandler creates a thrift handler for the cadence admin service
+// NewAdminHandler creates a thrift service for the cadence admin service
 func NewAdminHandler(
-	resource resource.Resource,
+	service Service,
```
Comment (davidporter-id-au) [inline service/frontend/adminHandler.go:150]: I don't like this much, passing itself as a parameter, but I don't really want to make this a huge refactor and this isn't really significantly changing much

File: common/resource/resourceImpl.go [hunk 2]
```diff
@@ -563,28 +577,71 @@ func (h *Impl) GetDispatcher() *yarpc.Dispatcher {
...
+	}
+	return cfgStoreClient
+}
+
 // Use the provided IsolationGroupStateHandler or the default one
-func ensureIsolationGroupStateHandlerOrDefault(
+// due to the config store being only available for some
+// persistence layers, *both* the configStoreClient and IsolationGroupState
+// will be optionally available
+func createIsolationGroupStateHandlerOrDefault(
```
Comment (davidporter-id-au) [inline common/resource/resourceImpl.go:624]: changing the name to `create...` to indicate that it's not guaranteed to not be nil. `ensure` has, in my experience mostly been used to do the opposite - to guarantee non-nil-ness

File: common/dynamicconfig/configstore/config_store_client.go [hunk 3]
```diff
@@ -342,6 +351,19 @@ func (csc *configStoreClient) ListValue(name dc.Key) ([]*types.DynamicConfigEntr
...
+	if !atomic.CompareAndSwapInt32(&csc.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	close(csc.doneCh)
+}
+
+func (csc *configStoreClient) Start() {
+	if !atomic.CompareAndSwapInt32(&csc.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
```
Comment (Shaddoll) [inline common/dynamicconfig/configstore/config_store_client.go:364]: nit: it seems better if we move `startUpdate` call from `new` method to `Start`

File: common/isolationgroup/defaultisolationgroupstate/state.go [hunk 4]
```diff
@@ -20,73 +20,41 @@
...
-	AllIsolationGroups []string
-}
-
 type defaultIsolationGroupStateHandler struct {
 	status                     int32
 	done                       chan struct{}
 	log                        log.Logger
 	domainCache                cache.DomainCache
 	globalIsolationGroupDrains dynamicconfig.Client
+	domainManager              persistence.DomainManager
```
Comment (Shaddoll) [inline common/isolationgroup/defaultisolationgroupstate/state.go:49]: Do we need this field?

Comment (davidporter-id-au) [reply common/isolationgroup/defaultisolationgroupstate/state.go:49]: no, this is obviated, good catch

File: common/resource/resourceImpl.go [hunk 5]
```diff
@@ -563,28 +577,71 @@ func (h *Impl) GetDispatcher() *yarpc.Dispatcher {
...
+	}
+	cscConfig := &csc.ClientConfig{
+		PollInterval:        dc.GetDurationProperty(dynamicconfig.IsolationGroupStateRefreshInterval)(),
+		UpdateRetryAttempts: dc.GetIntProperty(dynamicconfig.IsolationGroupStateUpdateRetryAttempts)(),
+		FetchTimeout:        dc.GetDurationProperty(dynamicconfig.IsolationGroupStateFetchTimeout)(),
+		UpdateTimeout:       dc.GetDurationProperty(dynamicconfig.IsolationGroupStateUpdateTimeout)(),
+	}
+	cfgStoreClient, err := configstore.NewConfigStoreClient(cscConfig, &params.PersistenceConfig, params.Logger, persistence.GlobalIsolationGroupConfig)
+	if err != nil {
+		// not possible to create the client under some persistence configurations, so this is expected
```
Comment (Shaddoll) [inline common/resource/resourceImpl.go:614]: nit: add a log

File: service/frontend/adminHandler_test.go [hunk 6]
```diff
@@ -109,7 +110,11 @@ func (s *adminHandlerSuite) SetupTest() {
 		EnableAdminProtection:  dynamicconfig.GetBoolPropertyFn(false),
 		EnableGracefulFailover: dynamicconfig.GetBoolPropertyFn(false),
 	}
-	s.handler = NewAdminHandler(s.mockResource, params, config).(*adminHandlerImpl)
+
+	ctrl := gomock.NewController(s.T())
+	dh := domain.NewMockHandler(ctrl)
```
Comment (Shaddoll) [inline service/frontend/adminHandler_test.go:115]: the parameter should be `s.controller`

File: service/frontend/workflowHandler_test.go [hunk 7]
```diff
@@ -119,6 +120,19 @@ func (s *workflowHandlerSuite) SetupTest() {
...
 	s.mockVersionChecker = client.NewMockVersionChecker(s.controller)
 
+	// these tests don't mock the domain handler
+	config := s.newConfig(dc.NewInMemoryClient())
+	s.domainHandler = domain.NewHandler(
+		config.domainConfig,
+		s.mockResource.GetLogger(),
+		s.mockResource.GetDomainManager(),
+		s.mockResource.GetClusterMetadata(),
+		domain.NewDomainReplicator(nil, s.mockResource.GetLogger()),
```
Comment (Shaddoll) [inline service/frontend/workflowHandler_test.go:130]: nil should be `s.mockProducer` and then you don't need the code change in `TestUpdateDomain_Success_FailOver`

File: common/isolationgroup/isolationgroupapi/handler.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package isolationgroupapi
+
+import (
+	"github.com/uber/cadence/common/domain"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+)
+
+// The MIT License (MIT)
```
Comment (Shaddoll) [inline common/isolationgroup/isolationgroupapi/handler.go:9]: nit: license should be moved to the top

File: common/isolationgroup/isolationgroupapi/interface.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination isolation_handler_mock.go -self_package github.com/uber/cadence/common/isolationgroup/isolationgrouphandler
+
+type Handler interface {
+	// GetGlobalState returns the current configuration of isolation-groups which apply to all domains
+	GetGlobalState(ctx context.Context) (*types.GetGlobalIsolationGroupsResponse, error)
+	// UpdateGlobalState updates isolation-groups which apply to all domains
+	UpdateGlobalState(ctx context.Context, state types.UpdateGlobalIsolationGroupsRequest) error
+	// GetDomainState is the read operation for getting the current state of a domain's isolation-groups
+	GetDomainState(ctx context.Context, request types.GetDomainIsolationGroupsRequest) (*types.GetDomainIsolationGroupsResponse, error)
```
Comment (Shaddoll) [inline common/isolationgroup/isolationgroupapi/interface.go:39]: It's a bit weird to pass the parameter by value not by pointer

File: service/frontend/adminHandler_test.go [hunk 10]
```diff
@@ -975,10 +900,10 @@ func Test_UpdateGlobalIsolationGroups(t *testing.T) {
...
 		Resource: &resource.Test{
-			Logger:          loggerimpl.NewNopLogger(),
-			IsolationGroups: nil, // valid state, the isolation-groups feature is not available for all persistence types
-			MetricsClient:   metrics.NewNoopMetricsClient(),
+			Logger:        loggerimpl.NewNopLogger(),
+			MetricsClient: metrics.NewNoopMetricsClient(),
 		},
+		isolationGroups: nil, // valid state, the isolation-groups feature is not available for all persistence types
 	}
 	assert.NotPanics(t, func() {
```
Comment (Shaddoll) [inline service/frontend/adminHandler_test.go:908]: it's probably better to check the error type, which implies that it doesn't panic

File: common/isolationgroup/isolationgroupapi/global-api-handlers_test.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	dynamicConfigResponse := []interface{}{}
+	json.Unmarshal(validCfgData, &dynamicConfigResponse)
+
+	tests := map[string]struct {
+		in           types.GetGlobalIsolationGroupsRequest
+		dcAffordance func(client *dynamicconfig.MockClient)
+		expected     *types.GetGlobalIsolationGroupsResponse
+		expectedErr  error
+	}{
+		"updating normal value": {
```
Comment (Shaddoll) [inline common/isolationgroup/isolationgroupapi/global-api-handlers_test.go:99]: nit: the message needs to be updated

File: common/isolationgroup/isolationgroupapi/global-api-handlers_test.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+				"zone-1": {Name: "zone-1", State: types.IsolationGroupStateHealthy},
+				"zone-2": {Name: "zone-2", State: types.IsolationGroupStateDrained},
+			}},
+			dcAffordance: func(client *dynamicconfig.MockClient) {
+				client.EXPECT().UpdateValue(
+					dynamicconfig.DefaultIsolationGroupConfigStoreManagerGlobalMapping,
+					gomock.Any(), // covering the mapping in the mapper unit-test instead
+				)
+			},
+		},
```
Comment (Shaddoll) [inline common/isolationgroup/isolationgroupapi/global-api-handlers_test.go:57]: nit: an error case

File: service/frontend/workflowHandler_test.go (hunk 7 above)
Comment (davidporter-id-au) [reply service/frontend/workflowHandler_test.go:130]: I do for a different reason, but good point, thank you

File: service/frontend/adminHandler_test.go (hunk 6 above)
Comment (davidporter-id-au) [reply service/frontend/adminHandler_test.go:115]: 🤦

File: common/isolationgroup/isolationgroupapi/interface.go (hunk 9 above)
Comment (davidporter-id-au) [reply common/isolationgroup/isolationgroupapi/interface.go:39]: this is a [go/glue convention](https://sourcegraph.uberinternal.com/code.uber.internal/go-code/-/blob/src/code.uber.internal/crack/luna-gateway/controller/benefits/benefits.go#L27) which isn't mine: Pass in values, return pointers to denote optionality, handle mapping, error handling, authorization at the outer layer and pass in validated values for the inner business logic. I don't feel strongly about it though, and in this instance it doesn't make a lot of difference. 

Obviously it's slightly more memory usage, but for operations like this it hardly matters.

File: common/isolationgroup/isolationgroupapi/global-api-handlers_test.go (hunk 11 above)
Comment (davidporter-id-au) [reply common/isolationgroup/isolationgroupapi/global-api-handlers_test.go:99]: not sure I understand?

File: common/isolationgroup/isolationgroupapi/global-api-handlers_test.go (hunk 12 above)
Comment (davidporter-id-au) [reply common/isolationgroup/isolationgroupapi/global-api-handlers_test.go:57]: tricky... I am not sure I can trigger an error case given the type system and the only thing that can trigger an error is the json serialization

**Review by Shaddoll — APPROVED** [review]

**Pull request merged**